src/VisualStudio/Core/Def/Experimentation/EnhancedColorExperiment.cs

```diff
@@ -0,0 +1,335 @@
+using System.Linq;
+using System.Runtime.InteropServices;
+using EnvDTE;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.VisualStudio.LanguageServices.Experimentation;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Settings;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using Microsoft.Win32;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Experimentation
+{
+    [Export(typeof(IWpfTextViewConnectionListener))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Analyzable)]
+    internal class EnhancedColorExperiment : IWpfTextViewConnectionListener
+    {
+        private const string UseEnhancedColorsFlight = "UseEnhancedColors";
+        private const string StopEnhancedColorsFlight = "StopEnhancedColors";
+        private const string UseEnhancedColorsSetting = "WindowManagement.Options.UseEnhancedColorsForManagedLanguages";
+
+        private readonly ISettingsManager _settingsManager;
+        private readonly EnhancedColorApplier _colorApplier;
+
+        private readonly bool _inUseEnhancedColorsFlight;
+        private readonly bool _inStopEnhancedColorsFlight;
+
+        private bool _done;
+
+        [ImportingConstructor]
+        private EnhancedColorExperiment([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, ExperimentationService experimentationService)
+        {
+            _inUseEnhancedColorsFlight = experimentationService?.IsExperimentEnabled(UseEnhancedColorsFlight) ?? false;
+            _inStopEnhancedColorsFlight = experimentationService?.IsExperimentEnabled(StopEnhancedColorsFlight) ?? false;
+
+            _colorApplier = new EnhancedColorApplier(serviceProvider);
+
+            _settingsManager = (ISettingsManager)serviceProvider.GetService(typeof(SVsSettingsPersistenceManager));
+
+            // Do not hook settings changed if we have stopped the experiment
+            if (_inStopEnhancedColorsFlight)
+            {
+                return;
+            }
+
+            // We need to update the theme whenever the Preview Setting changes or the VS Theme changes.
+            _settingsManager.GetSubset(UseEnhancedColorsSetting).SettingChangedAsync += UseEnhancedColorsSettingChangedAsync;
+            VSColorTheme.ThemeChanged += VSColorTheme_ThemeChanged;
+        }
+
+        private void VSColorTheme_ThemeChanged(ThemeChangedEventArgs e)
+        {
+            VsTaskLibraryHelper.CreateAndStartTask(VsTaskLibraryHelper.ServiceInstance, VsTaskRunContext.UIThreadIdlePriority, UpdateThemeColors);
+        }
+
+        private Task UseEnhancedColorsSettingChangedAsync(object sender, PropertyChangedEventArgs args)
+        {
+            return Task.Run(() => UpdateThemeColors());
```

Comment:
Red flag that you're kicking off work that's using DTE on a background thread in a fire and forget way.